src/main/java/org/cryptomator/cloudaccess/MetadataRequestAggregator.java

```diff
@@ -95,68 +101,63 @@ public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEn
-						quotaCache.invalidateAll();
-					} else {
-						itemMetadataCache.invalidate(file);
-					}
-				});
+		return delegate.write(file, replace, data, size, lastModified, progressListener).whenComplete((nullReturn, exception) -> {
+			if (exception != null) {
+				invalidateAggregators(file);
+			}
+		});
 	}
 
 	@Override
 	public CompletionStage<CloudPath> createFolder(CloudPath folder) {
-		return delegate.createFolder(folder) //
-				.whenComplete((metadata, exception) -> {
-					itemMetadataCache.invalidate(folder);
-					quotaCache.invalidateAll();
-				});
+		return delegate.createFolder(folder).whenComplete((metadata, exception) -> invalidateAggregators(folder));
 	}
 
 	@Override
 	public CompletionStage<Void> deleteFile(CloudPath file) {
-		return delegate.deleteFile(file) //
-				.whenComplete((nullReturn, exception) -> {
-					CompletionStage<CloudItemMetadata> future = CompletableFuture.failedFuture(new NotFoundException());
-					itemMetadataCache.put(file, future);
-					quotaCache.invalidateAll();
-				});
+		return delegate.deleteFile(file);
 	}
 
 	@Override
 	public CompletionStage<Void> deleteFolder(CloudPath folder) {
-		return delegate.deleteFolder(folder) //
-				.whenComplete((nullReturn, exception) -> {
-					evictIncludingDescendants(folder);
-					CompletionStage<CloudItemMetadata> future = CompletableFuture.failedFuture(new NotFoundException());
-					itemMetadataCache.put(folder, future);
-					quotaCache.invalidateAll();
-				});
+		return delegate.deleteFolder(folder).whenComplete((nullReturn, exception) -> evictIncludingDescendants(folder));
 	}
 
 	@Override
 	public CompletionStage<CloudPath> move(CloudPath source, CloudPath target, boolean replace) {
-		return delegate.move(source, target, replace) //
-				.whenComplete((path, exception) -> {
-					evictIncludingDescendants(source);
-					evictIncludingDescendants(target);
-					quotaCache.invalidateAll();
-				});
+		return delegate.move(source, target, replace).whenComplete((path, exception) -> {
+			evictIncludingDescendants(source);
+			evictIncludingDescendants(target);
+		});
 	}
 
 	private void evictIncludingDescendants(CloudPath cleartextPath) {
-		for (var path : itemMetadataCache.asMap().keySet()) {
+		for (var path : itemMetadataRequestAggregator.asMap().keySet()) {
 			if (path.startsWith(cleartextPath)) {
-				itemMetadataCache.invalidate(path);
+				itemMetadataRequestAggregator.invalidate(path);
+			}
+		}
+		for (var path : cloudItemListRequestAggregator.asMap().keySet()) {
+			if (path.getKey().startsWith(cleartextPath)) {
+				cloudItemListRequestAggregator.invalidate(path);
+			}
+		}
+	}
+
+	private void invalidateAggregators(CloudPath cleartextPath) {
+		itemMetadataRequestAggregator.invalidate(cleartextPath);
+
+		for (var path : cloudItemListRequestAggregator.asMap().keySet()) {
+			if (path.getKey().equals(cleartextPath)) {
+				cloudItemListRequestAggregator.invalidate(path);
+				return;
 			}
 		}
```

Comment:
what is this loop supposed to do? looks to me as if it only invalidates the first key that matches the given path. why not just use `invalidate(cleartextPath)` directly?

Reply:
The return was a faulty optimization. As the key of the `cloudItemListRequestAggregator` is a Map containing the `Path` and `Pagetoken`, we need to iterate over all Paths and remove those entries matching this path.